File: llm_generated_code/ChatGPT/Python/Level1/Q4.py
def max_profit(prices):
    if not prices:
        return 0

    # Initialize min_price to a very high value and max_profit to 0
    min_price = float('inf')
    max_profit = 0

    # Traverse through the list of prices
    for price in prices:
        if price < min_price:
            # Update min_price to the current price if it's lower than previously found
            min_price = price
        elif price - min_price > max_profit:
            # Calculate the profit from the current price minus the minimum price found so far
            max_profit = price - min_price

    return max_profit

File: llm_generated_code/ChatGPT/CPP/Level1/Q4.cpp

#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int maxProfit(const vector<int>& prices) {
    if (prices.empty()) return 0;

    int minPrice = prices[0];
    int maxProfit = 0;

    for (int i = 1; i < prices.size(); i++) {
        // Update minimum price if current price is lower
        minPrice = min(minPrice, prices[i]);

        // Calculate the profit if we sell at the current price
        int profit = prices[i] - minPrice;

        // Update maximum profit if this profit is higher
        maxProfit = max(maxProfit, profit);
    }

    return maxProfit;
}

int main() {
    vector<int> prices = {7, 1, 5, 3, 6, 4}; // Example array
    int result = maxProfit(prices);
    cout << "Maximum profit: " << result << endl;
    return 0;
}
